### PR TITLE
Newsletter Feature

### PR DESCRIPTION
**Pull request by theBeginner86**
## What is the change?
- Files Added:

         .env example:
                 Added all the variable which is required by the Google API for authentication and sending mail from Alphavio's gmail account. You must copy all the values into it to .env file. Also, we can decide if we want to give these same values to all the developers, so that their setup time is reduced or we may add the generation process for those values into the documentation. For security reasons, I have not included them in the PR but dm you on discord.
        middlewares/newsletter.js:
                 This the main newsletter file which handles majority of the newsletter work. Here Google APIs are initialised,  the message to be sent is composed dynamically, all the recipients are found and then a mail is sent to their respective emails with a blog having max views. This is sent only when there exists one or more blogs.
        views/newsletter.html:
                  This is the html file which is rendered on the fly with dynamically added variables of the blog.
        views/useritems/newsletter.ejs:
                   This file is displayed onto the dashboard of the user and gives the user a choice to subscribe or unsubscribe the newsletter. This rendering of this files depends upon whether the user has initially subscribed or not.

- Files Appened:

        app.js:
                  Here, I have scheduled the the function sendNewsletter to be executed at the 1st day of each month using node-cron. But there is a comment which could help you to change it to every 15 secs. This might help you in the reviewing process
        models/User.model.js:
                  Here, I have added a new attribute, ie, subscriptionStatus with a value type of boolean. So, if user has subscribed to the newsletter then it holds true otherwise it holds false. By default, it would hold the false value.
        package-lock.json package.json:
                  Here, I have added three node modules: googleapis, node-cron, handlebars. This was important to add this feature.
        routes/user.router.js:
                 Here, I have added two post request routes. One is used when the user wants to subscribe the newsletter and then it changes the subscriptionStatus to true and the other is used when the user wants to unsubscribe the newsletter and then it changes the subscriptionStatus to false. This basically helps the sendNewsletter function to determine to determine the recipient of the mail with blogs.
        views/useritems/dashboard.ejs:
                  Here, the newsletter.ejs file is rendered. This helps the user to keep track of whether he/she has subscribed to the newsletter or not.
## Related issue?
close: #276 

## How was it tested?
Tested Locally

## Checklist:
Before you create this PR, confirm all the requirements listed below by checking the checkboxes `[x]`:

- [x] Have you followed the [Contribution Guidelines](https://github.com/ALPHAVIO/BlogSite/blob/master/CONTRIBUTING.md) while contributing.
- [x] Have you checked there aren't other open [Pull Requests](https://github.com/ALPHAVIO/BlogSite/pulls) for the same update/change?
- [x] Did you lint your code before making the Pull Request?
- [] Have you made corresponding changes to the documentation?
- [x] Your submission doesn't break any existing feature.
- [x] Have you tested the code before submission?

## Screenshots or Video:

The mails received by the recipients is as follows:

Screenshot 2021-05-24 at 12 28 11 AM

Screenshot 2021-05-24 at 12 28 02 AM

I know the css is not that great but we can solve it in subsequent issues.

The dashboard looks like:

Screenshot 2021-05-24 at 1 19 49 AM

Screenshot 2021-05-24 at 1 20 01 AM

### CODE REVIEW

**Comment by Pranav016** [comment]
Can you add a demo video about how the whole feature is working, alot of changes in this single PR. I would still suggest breaking it down to increase the points

**Comment by Pranav016** [comment]
> Can you add a demo video about how the whole feature is working, alot of changes in this single PR. I would still suggest breaking it down to increase the points

add a demo to get the PR merge asap before gssoc deadline

**Comment by theBeginner86** [comment]
> > Can you add a demo video about how the whole feature is working, alot of changes in this single PR. I would still suggest breaking it down to increase the points
> 
> add a demo to get the PR merge asap before gssoc deadline

Yeah, I was making that only. Now uploading on google drive. Will send the link here in about 15 mins.

**Comment by theBeginner86** [comment]
@Pranav016 
The following includes the google drive link where I have added the demo video. 
Sorry, for making this video a little long. I have tried to explain each and everything properly.

https://drive.google.com/drive/folders/1Mj1_5NY6rLIRn2DpUDHRB9Xap_UuaDaN?usp=sharing

Let me know your views on it.

**Comment by theBeginner86** [comment]
One important thing is that, do add the .env variables that I shared with you on discord, in your cloned repo, so that you can also run the project on your machine. Also, we need to add documentation for setting up the project, if we decide not to share Alphavio's gmail credentials with other developers.

**Comment by RushanKhan1** [comment]
> @Pranav016
> The following includes the google drive link where I have added the demo video.
> Sorry, for making this video a little long. I have tried to explain each and everything properly.
> 
> https://drive.google.com/drive/folders/1Mj1_5NY6rLIRn2DpUDHRB9Xap_UuaDaN?usp=sharing
> 
> Let me know your views on it.

@Pranav016  I watched the video, and the whole thing seems functional to me.

@theBeginner86 I really appreciate the effort that you put into making the video but I would recommend that if you are demonstrating something here or elsewhere it is a good practice to try to keep it as short and concise as possible because open source maintainers have limited time and they might not give your pr a look if it requires a big time investment.

File: routes/user.router.js [hunk 1]
```diff
@@ -394,4 +394,36 @@ router.delete('/profile/:userId/delete', auth, async (req, res) => {
...
+			await user.updateOne({ subscriptionStatus: false });
+			return res.redirect('/dashboard');
+		} catch (error) {
+			return res.redirect('/error');
+		}
+	} catch (error) {
+		return res.redirect('/error');
+	}
+});
+
```
Comment (Pranav016) [inline routes/user.router.js:428]: make a separate router for newsletter

File: middlewares/newsletter.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+const composeMessage = async () => {
+	const {
+		blogTitle,
+		blogContent,
+		category,
+		author,
+		timestamps,
+		noOfViews,
+		likes,
+	} = await Blog.findOne({ status: 'Public' }).sort({ noOfViews: -1 });
```
Comment (Pranav016) [inline middlewares/newsletter.js:55]: did not understand this part ` noOfViews: -1 `

**Review by Pranav016 — COMMENTED** [review]
Great work!

File: middlewares/newsletter.js (hunk 2 above)
Comment (theBeginner86) [reply middlewares/newsletter.js:55]: Actually, here -1 tells the mongodb to sort the data in descending order according to noOfViews values. Then the first one with public status will be returned. So, this value will have the highest no of views and with the status of public.

File: routes/user.router.js (hunk 1 above)
Comment (theBeginner86) [reply routes/user.router.js:428]: Like should I do it now or is it fine. I am asking this because you merged the PR.

Comment (Pranav016) [reply routes/user.router.js:428]: You can make a separate PR.

Comment (theBeginner86) [reply routes/user.router.js:428]: ok. just one more doubt. Here by separate router, you mean the route should be like: localhost:3000/newsletter/subscribe/ 
and
local localhost:3000/newsletter/unsubscribe/ 
or it's something else you are asking?

Comment (Pranav016) [reply routes/user.router.js:428]: > ok. just one more doubt. Here by separate router, you mean the route should be like: localhost:3000/newsletter/subscribe/
> and
> localhost:3000/newsletter/unsubscribe/
> or its the separate file you are talking about? like routes/newsletter.router.js

yes it is currently present in the user.router file, I want you to make a new file for the newsletter router to put these subscribe and unsubscribe routes and just do `app.use('/newsletter', <router-name>)`  in index file

Comment (theBeginner86) [reply routes/user.router.js:428]: ok I understood it. Will make a PR now.

**Pull request merged**